### PR TITLE
Added RegisterVocabulary event

### PR DESCRIPTION
**Pull request by chevtche**

### CODE REVIEW

**Comment by chevtche** [comment]
Retest this please

**Comment by chevtche** [comment]
Retest this please

**Comment by chevtche** [comment]
Retest this please

**Comment by chevtche** [comment]
Retest this please

**Comment by chevtche** [comment]
Retest this please

**Comment by chevtche** [comment]
Retest this please

**Comment by chevtche** [comment]
Retest this please

File: zeq/detail/registerVocabulary.fbs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+// Copyright (c) 2015, Human Brain Project
+//                     Grigori Chevtchenko <grigori.chevtchenko@epfl.ch>
+
+namespace zeq.vocabulary;
+
+table RegisterVocabulary
+{
+  restNames:[string];
+  eventLows:[ulong];
+  eventHighs:[ulong];
```
Comment (eile) [inline zeq/detail/registerVocabulary.fbs]: swap these two, unless your brain is little endian. ;)

File: zeq/detail/registerVocabulary.fbs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+// Copyright (c) 2015, Human Brain Project
+//                     Grigori Chevtchenko <grigori.chevtchenko@epfl.ch>
+
+namespace zeq.vocabulary;
+
+table RegisterVocabulary
```
Comment (eile) [inline zeq/detail/registerVocabulary.fbs]: Call it only Vocabulary?

File: zeq/detail/registerVocabulary.fbs [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+{
+  restNames:[string];
+  eventLows:[ulong];
+  eventHighs:[ulong];
+  schemas:[string];
+}
```
Comment (eile) [inline zeq/detail/registerVocabulary.fbs]: Actually, shouldn't these be a vector of structs (struct: all data for one event)?

Comment (chevtche) [inline zeq/detail/registerVocabulary.fbs]: Yes, but flattbuffers does not support strings in structures LOL.

**Comment by chevtche** [comment]
Retest this please

File: zeq/detail/registerVocabulary.fbs (hunk 3 above)
Comment (eile) [inline zeq/detail/registerVocabulary.fbs]: Ok, then it's ok as is. Maybe add a comment.

File: zeq/detail/vocabulary.cpp [hunk 4]
```diff
@@ -26,6 +28,57 @@ stde::hash_map< uint128_t, std::string >& getRegistry()
...
+    std::vector< uint64_t > eventTypeLows;
+    std::vector< uint64_t > eventTypeHighs;
+    std::vector< flatbuffers::Offset<flatbuffers::String> > schemas;
+
+    for ( EventDescriptorsVector::const_iterator it = vocabulary.begin();
+          it!=vocabulary.end(); ++it )
+    {
+        restNames.push_back( fbb.CreateString( it->getRestName() ) );
+        eventTypeLows.push_back( it->getEventType().low() );
+        eventTypeHighs.push_back( it->getEventType().high() );
```
Comment (eile) [inline zeq/detail/vocabulary.cpp]: here too.

File: zeq/detail/vocabulary.cpp [hunk 5]
```diff
@@ -26,6 +28,57 @@ stde::hash_map< uint128_t, std::string >& getRegistry()
...
+        schemas.push_back( fbb.CreateString( it->getSchema() ) );
+    }
+
+    auto restNamesForZeq = fbb.CreateVector( &restNames[0], restNames.size() );
+    auto eventTypeLowsForZeq = fbb.CreateVector( &eventTypeLows[0], eventTypeLows.size() );
+    auto eventTypeHighsForZeq = fbb.CreateVector( &eventTypeHighs[0], eventTypeHighs.size() );
```
Comment (eile) [inline zeq/detail/vocabulary.cpp]: ...

File: zeq/types.h [hunk 6]
```diff
@@ -32,7 +32,9 @@ namespace zeq
 class Event;
 class Publisher;
 class Subscriber;
+struct EventDescriptor;
 
+typedef std::vector< ::zeq::EventDescriptor > EventDescriptorsVector;
```
Comment (eile) [inline zeq/types.h]: Our common naming patter is 'EventDescriptors' for vector-of-type

Comment (chevtche) [inline zeq/types.h]: Ok I will change that but I think, in general, EventDescriptorsVector is a more clear name.

**Comment by chevtche** [comment]
Retest this please

**Comment by chevtche** [comment]
Retest this please

**Comment by chevtche** [comment]
Can this be merged?

File: zeq/detail/registerVocabulary.fbs [hunk 7]
```diff
@@ -0,0 +1,16 @@
+// Copyright (c) 2015, Human Brain Project
```
Comment (tribal-tec) [inline zeq/detail/registerVocabulary.fbs]: rename file to vocabulary.fbs. Pay attention to the _generated files as well

File: zeq/detail/vocabulary.cpp [hunk 8]
```diff
@@ -26,6 +28,57 @@ stde::hash_map< uint128_t, std::string >& getRegistry()
...
+{
+    ::zeq::Event event( EVENT_VOCABULARY );
+    flatbuffers::FlatBufferBuilder& fbb = event.getFBB();
+
+    std::vector< flatbuffers::Offset<flatbuffers::String> > restNames;
+    std::vector< uint64_t > eventTypeLows;
+    std::vector< uint64_t > eventTypeHighs;
+    std::vector< flatbuffers::Offset<flatbuffers::String> > schemas;
+
+    for ( EventDescriptors::const_iterator it = vocabulary.begin();
```
Comment (tribal-tec) [inline zeq/detail/vocabulary.cpp]: BOOST_FOREACH

File: zeq/detail/vocabulary.cpp [hunk 9]
```diff
@@ -26,6 +28,57 @@ stde::hash_map< uint128_t, std::string >& getRegistry()
...
+
+    fbb.Finish( builder.Finish() );
+    return event;
+}
+
+EventDescriptors deserializeVocabulary( const Event& event )
+{
+    auto data = GetVocabulary( event.getData() );
+
+    std::vector< ::zeq::EventDescriptor > vocabulary;
```
Comment (tribal-tec) [inline zeq/detail/vocabulary.cpp]: reserve to data->restNames().size()?

File: zeq/eventDescriptor.h [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+struct EventDescriptor : public boost::noncopyable
+{
+    EventDescriptor( const std::string& restName, const lunchbox::uint128_t& eventType,
+                     const std::string& schema );
+
+    EventDescriptor( EventDescriptor&& rhs );
+
+    ~EventDescriptor();
+
+    const std::string& getRestName() const;
```
Comment (tribal-tec) [inline zeq/eventDescriptor.h]: doxygen please

File: zeq/eventDescriptor.h [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    const lunchbox::uint128_t& getEventType() const;
+    const std::string& getSchema() const;
+private:
+
+    EventDescriptor& operator=( EventDescriptor&& rhs );
```
Comment (tribal-tec) [inline zeq/eventDescriptor.h]: shouldn't be needed if derived from boost::noncopyable

File: zeq/vocabulary.h [hunk 12]
```diff
@@ -28,6 +28,21 @@ namespace vocabulary
 ZEQ_API Event serializeEcho( const std::string& message );
 ZEQ_API std::string deserializeEcho( const Event& event );
 
+/**
+ * Serialize the given vocabulary into an Event of type EVENT_REGISTERVOCABULARY.
```
Comment (tribal-tec) [inline zeq/vocabulary.h]: typo EVENT_REGISTERVOCABULARY

File: zeq/vocabulary.h [hunk 13]
```diff
@@ -28,6 +28,21 @@ namespace vocabulary
...
+ * @param vocabulary the vocabulary ( a vector of CommandDescriptors ).
+ * @return the serialized event.
+ */
+ZEQ_API Event serializeVocabulary( const EventDescriptors& vocabulary );
+
+/**
+ * Deserialize the given EVENT_REGISTERVOCABULARY event into a vector of CommandDescriptors.
```
Comment (tribal-tec) [inline zeq/vocabulary.h]: dito

File: zeq/vocabulary.h [hunk 14]
```diff
@@ -28,6 +28,21 @@ namespace vocabulary
...
+ * @param event the zeq EVENT_REGISTERVOCABULARY.
```
Comment (tribal-tec) [inline zeq/vocabulary.h]: dito

File: zeq/types.h [hunk 15]
```diff
@@ -32,7 +32,9 @@ namespace zeq
 class Event;
 class Publisher;
 class Subscriber;
+struct EventDescriptor;
 
+typedef std::vector< ::zeq::EventDescriptor > EventDescriptors;
```
Comment (tribal-tec) [inline zeq/types.h]: you should not need ::zeq:: here

**Comment by eile** [comment]
So far there are still unaddressed comments.

File: zeq/vocabulary.h (hunk 12 above)
Comment (eile) [inline zeq/vocabulary.h]: Plus event is now name EVENT_VOCABULARY

File: zeq/eventDescriptor.h (hunk 11 above)
Comment (chevtche) [inline zeq/eventDescriptor.h]: It is move not copy.

**Comment by chevtche** [comment]
Retest this please

**Comment by chevtche** [comment]
Retest this please

**Comment by chevtche** [comment]
All comments addressed. Can this be merged?

File: zeq/detail/vocabulary.cpp (hunk 5 above)
Comment (eile) [inline zeq/detail/vocabulary.cpp]: Here are still two instance where you serialize low before high.

File: zeq/detail/vocabulary.cpp [hunk 16]
```diff
@@ -26,6 +30,57 @@ stde::hash_map< uint128_t, std::string >& getRegistry()
...
+
+    auto restNamesForZeq = fbb.CreateVector( &restNames[0], restNames.size() );
+    auto eventTypeLowsForZeq = fbb.CreateVector( &eventTypeLows[0], eventTypeLows.size() );
+    auto eventTypeHighsForZeq = fbb.CreateVector( &eventTypeHighs[0], eventTypeHighs.size() );
+    auto schemasForZeq = fbb.CreateVector( &schemas[0], schemas.size() );
+
+    VocabularyBuilder builder( fbb );
+    builder.add_restNames( restNamesForZeq );
+    builder.add_eventLows( eventTypeLowsForZeq );
+    builder.add_eventHighs( eventTypeHighsForZeq );
```
Comment (eile) [inline zeq/detail/vocabulary.cpp]: three.

File: zeq/detail/vocabulary.cpp (hunk 9 above)
Comment (eile) [inline zeq/detail/vocabulary.cpp]: First rule of optimization?

**Comment by eile** [comment]
+1 otherwise

**Comment by chevtche** [comment]
Retest this please

**Comment by chevtche** [comment]
Retest this please

File: zeq/CMakeLists.txt [hunk 17]
```diff
@@ -18,13 +18,15 @@ set(ZEQ_PUBLIC_HEADERS
   receiver.h
   subscriber.h
   types.h
-  vocabulary.h)
+  vocabulary.h
+  eventDescriptor.h)
```
Comment (tribal-tec) [inline zeq/CMakeLists.txt]: alphabetic order please

File: zeq/CMakeLists.txt [hunk 18]
```diff
@@ -18,13 +18,15 @@ set(ZEQ_PUBLIC_HEADERS
...
 
 set(ZEQ_HEADERS
   detail/broker.h
   detail/event.h
   detail/socket.h
-  detail/vocabulary.h)
+  detail/vocabulary.h
+  detail/eventDescriptor.h)
```
Comment (tribal-tec) [inline zeq/CMakeLists.txt]: alphabetic order please

File: zeq/CMakeLists.txt [hunk 19]
```diff
@@ -35,7 +37,8 @@ set(ZEQ_SOURCES
   publisher.cpp
   receiver.cpp
   subscriber.cpp
-  vocabulary.cpp)
+  vocabulary.cpp
+  eventDescriptor.cpp)
```
Comment (tribal-tec) [inline zeq/CMakeLists.txt]: alphabetic order please

File: zeq/eventDescriptor.h [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+
+/**
+ * This structure holds informations about a command from some zeq vocabulary.
+ * It contains the REST api name, the 128bit zeq event id and the schema
+ * needed for automatic JSON serialization/deserialization.
+ */
+struct EventDescriptor : public boost::noncopyable
+{
+    /**
+     * Create am EventDescriptor.
```
Comment (tribal-tec) [inline zeq/eventDescriptor.h]: typo 'am'

File: zeq/eventDescriptor.h [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+     * @param restName the string used for REST command
+     * @param eventType the zeq event's uint128
+     * @param schema the flatbuffers schema as string
+     */
+    ZEQ_API EventDescriptor( const std::string& restName, const lunchbox::uint128_t& eventType,
+                     const std::string& schema );
```
Comment (tribal-tec) [inline zeq/eventDescriptor.h]: indentation

File: zeq/vocabulary.h [hunk 22]
```diff
@@ -28,6 +28,21 @@ namespace vocabulary
 ZEQ_API Event serializeEcho( const std::string& message );
 ZEQ_API std::string deserializeEcho( const Event& event );
 
+/**
+ * Serialize the given vocabulary into an Event of type EVENT_VOCABULARY.
+ * @param vocabulary the vocabulary ( a vector of CommandDescriptors ).
```
Comment (tribal-tec) [inline zeq/vocabulary.h]: CommandDescriptor???

File: zeq/vocabulary.h [hunk 23]
```diff
@@ -28,6 +28,21 @@ namespace vocabulary
...
+ * @return the serialized event.
+ */
+ZEQ_API Event serializeVocabulary( const EventDescriptors& vocabulary );
+
+/**
+ * Deserialize the given EVENT_VOCABULARY event into a vector of CommandDescriptors.
```
Comment (tribal-tec) [inline zeq/vocabulary.h]: dito?

File: zeq/vocabulary.h [hunk 24]
```diff
@@ -28,6 +28,21 @@ namespace vocabulary
...
+ * @param event the zeq EVENT_VOCABULARY.
+ * @return the vector of CommandDescriptors.
```
Comment (tribal-tec) [inline zeq/vocabulary.h]: dito?

File: zeq/detail/vocabulary.cpp (hunk 9 above)
Comment (chevtche) [inline zeq/detail/vocabulary.cpp]: It is done: vocabulary.reserve( data->restNames()->Length() ); Is it wrong ?

**Comment by chevtche** [comment]
Retest this please

**Comment by chevtche** [comment]
Updated

File: zeq/eventDescriptor.h [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+
+#include <zeq/types.h>
+#include <boost/noncopyable.hpp>
+
+namespace zeq
+{
+namespace detail { struct EventDescriptor; }
+
+/**
+ * This structure holds informations about a command from some zeq vocabulary.
```
Comment (tribal-tec) [inline zeq/eventDescriptor.h]: ...about an Event from a vocabulary.

**Comment by tribal-tec** [comment]
+1 otherwise. Thx for the changes!

**Comment by chevtche** [comment]
Retest this please

**Pull request merged**